### PR TITLE
[RFC] vim-patch:7.4.503

### PR DESCRIPTION
**Pull request by iperry**
Problem:    Cannot append a list of lines to a file.
Solution:   Add the append option to writefile(). (Yasuhiro Matsumoto)

https://code.google.com/p/vim/source/detail?r=v7-4-503

Tested with: make oldtest, make test on OS X.

Signed-off-by: Perry Hung iperry@gmail.com

### CODE REVIEW

**Comment by iperry** [comment]
Hi, first time PR for neovim.

I actually started with the intention of throwing together a quick fix from upstream for #1999. 

Then I realized how much of a pain it was to port some of these upstream patches due to the way Bram breaks up his patches. (e.g (https://code.google.com/p/vim/source/detail?r=f654ad95fd4e25e014dda71963b9745a4bf5f83b#).

This  one is particular mixes totally unrelated documentation cleanups and TODO changes with an actual actual bug fix in syntax/c.vim. These documentation fixes impose a dependency on patches that may not have been ported yet, such as this pull request.

The current system seems to track upstream patch to nvim patch with a one-to-one mapping.  This makes it very difficult to read or prepare these PRS, since I have to touch a whole bunch of unrelated systems, resolve their dependencies, follow each of those dependencies all the way back to the branch point, and then make sure that each of those patches have been applied, before I can fix one simple bug that's already been fixed. It seems many of the runtime patches in particular are snapshots of Bram's tree every once in a while.

There must be a saner way of doing this? What is the recommended method doing this? Perhaps separately implement each feature in many different requests, or be much more aggressive about keeping current with upstream.

**Comment by iperry** [comment]
I'm not sure why the Travis build failed. I ran the same tests (oldtest, test) on GCC 4.8.2 on a Linux x86 system as well as a custom clang built from trunk with OS X and it also passed. 

Are the Travis CI builds known to be stable? Nothing I touched in this PR should have caused test 61, a totally unrelated test to fail.

**Comment by Pyrohh** [comment]
> Are the Travis CI builds known to be stable? Nothing I touched in this PR should have caused test 61, a totally unrelated test to fail.

test61 has failed in PRs I made only concerning documentation, so there's that.

**Comment by iperry** [comment]
Note for reviewers:

I left out the mapping, signs, and utf8 test because they have nothing to do with this feature.

**Comment by justinmk** [comment]
> The current system seems to track upstream patch to nvim patch with a one-to-one mapping. This makes it very difficult to read or prepare these PRS

That's not normal for most Vim patches (except the runtime updates, as you mentioned). 

> or be much more aggressive about keeping current with upstream.

Well, yeah. Merging earlier patches before later patches is generally encouraged, otherwise it will cause pain as you have noticed. Solution is to KISS and just merge the early patches first.

**Comment by justinmk** [comment]
> I left out the mapping, signs, and utf8 test because they have nothing to do with this feature.

I am assuming that adding those lines would fail the build. In that case I created https://github.com/neovim/neovim/issues/2185 as a reminder to add those tests to the Makefile.

File: runtime/doc/eval.txt [hunk 1]
```diff
@@ -6593,13 +6593,19 @@ winwidth({nr})						*winwidth()*
...
 		Write |List| {list} to file {fname}.  Each list item is
 		separated with a NL.  Each list item must be a String or
 		Number.
-		When {binary} is equal to "b" binary mode is used: There will
+		When {flags} contains "b" then binary mode is used: There will
 		not be a NL after the last list item.  An empty item at the
 		end does cause the last line in the file to end in a NL.
+
+		When {flags} contains "a" then append mode is used, lines are
+		append to the file: >
```
Comment (justinmk) [inline runtime/doc/eval.txt]: Just for future reference, it's ok to fix glaring grammar/spelling errors like this (append => appended).

File: src/nvim/testdir/Makefile [hunk 2]
```diff
@@ -37,7 +37,8 @@ SCRIPTS := test_autoformat_join.out                                    \
            test_qf_title.out                                           \
            test_breakindent.out                                        \
            test_insertcount.out                                        \
```
Comment (justinmk) [inline src/nvim/testdir/Makefile]: this line can be removed, `test_insertcount.out` was migrated to `test/functional/legacy`.

File: src/nvim/testdir/Makefile [hunk 3]
```diff
@@ -37,7 +37,8 @@ SCRIPTS := test_autoformat_join.out                                    \
...
-           test_utf8.out
+           test_utf8.out                                               \
```
Comment (justinmk) [inline src/nvim/testdir/Makefile]: this line can be removed, test_insertcount.out was migrated to test/functional/legacy.

**Comment by justinmk** [comment]
SInce you are new, maybe you don't know, but we are migrating all of the `testdir` tests to lua tests. You can see the tests in `test/functional/legacy/` for examples. Why not convert the test in this patch to a lua test? It's small, and should be easy.

**Comment by iperry** [comment]
Sure. I have no problem with that. Is there a pointer to how to do this, or an easy example to look at?

**Comment by justinmk** [comment]
Was looking for the issue. Here it is: https://github.com/neovim/neovim/issues/1286

**Comment by iperry** [comment]
I rebased the pull request:

1) The old legacy test has been converted to use the new framework
2) On @justinmk's request I removed test_utf8 and test_insertcount from the old testdir, though I would really prefer to do this in a separate commit to keep things logically separate.
3) spelling fixup s/append/appended

**Comment by justinmk** [comment]
merged, thanks!

**Pull request closed without merging**